### PR TITLE
feat: browser settings in the settings page

### PR DESCRIPTION
**Pull request by g-linville**
for #321

### CODE REVIEW

File: electron/config.mjs [hunk 1]
```diff
@@ -4,6 +4,25 @@ import { join, dirname, parse } from 'path';
...
+  } else if (os.platform() === 'win32') {
+    configDir = join(homeDir, 'AppData', 'Local');
+  } else if (os.platform() === 'linux') {
+    configDir = process.env.XDG_CONFIG_HOME || join(homeDir, '.config');
+  } else {
+    throw new Error('Unsupported platform');
+  }
+
+  return join(configDir, 'acorn', 'browsersettings.json');
+}
```
Comment (g-linville) [inline electron/config.mjs:25]: I hate how this function is duplicated from actions/browser.tsx. But I'm not sure if there is a way for me to define it in one place and have it be callable from both places. Please let me know if there is a way to do that.

File: electron/main.mjs [hunk 2]
```diff
@@ -31,6 +31,7 @@ async function startServer() {
   process.env.WORKSPACE_DIR = config.workspaceDir;
   process.env.GPTSCRIPT_GATEWAY_URL = config.gatewayUrl;
   process.env.GPTSCRIPT_OPENAPI_REVAMP = 'true';
+  process.env.GPTSCRIPT_BROWSER_SETTINGS_FILE = config.browserSettingsFile;
```
Comment (g-linville) [inline electron/main.mjs:34]: This environment variable is read by the browser tool itself so that it knows where to look for its settings.

**Review by ryanhopperlowe — APPROVED** [review]

File: electron/main.mjs (hunk 2 above)
Comment (tylerslaton) [reply electron/main.mjs:34]: Does this mean that for various tools that have settings that they will all have their own settings files?

File: actions/browser.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    const settings = fs.readFileSync(location, 'utf-8');
+    return JSON.parse(settings) as BrowserSettings;
+  }
+
+  return defaultSettings;
+}
+
+export async function setBrowserSettings(settings: BrowserSettings) {
+  const location = await browserSettingsLocation();
+  fs.writeFileSync(location, JSON.stringify(settings, null, 2));
```
Comment (tylerslaton) [inline actions/browser.tsx:29]: 👍🏻 Like the use of the formatting part of stringify.

File: app/settings/page.tsx [hunk 4]
```diff
@@ -22,6 +46,69 @@ export default function Settings() {
...
+                    setSettings((prevState) => ({
+                      ...prevState,
+                      useDefaultSession: isSelected,
+                    }));
+                  }}
+                >
+                  Use Default Session
+                </Switch>
+                <Tooltip
+                  content="Use Default Session will use your existing default Google Chrome session. When this setting is enabled, Chrome must be closed before running the browser tool."
```
Comment (tylerslaton) [inline app/settings/page.tsx:87]: nit:

<details>

<summary>Right now this goes all the way to the edge of my app and its a little hard to read.</summary>

Screenshot 2024-09-06 at 9 38 07 AM

</details>

```suggestion
                <Tooltip
                  classNames={{base: "w-1/2"}}
                  content="Use Default Session will use your existing default Google Chrome session. When this setting is enabled, Chrome must be closed before running the browser tool."
```

<details>

<summary>Could we add that max width so it looks more like this?</summary>

Screenshot 2024-09-06 at 9 43 40 AM

</details>

File: electron/config.mjs (hunk 1 above)
Comment (tylerslaton) [reply electron/config.mjs:25]: That I personally know of, no. I've had a lot of similar headaches. Ideally we'd make a global lib package that both the Next app and electron could import. To do that right though I think we'd need the socket server and config code to be in typescript so it isn't super simple in just this PR.

**Review by tylerslaton — APPROVED** [review]
Had a small nit I'd like to see implemented and some questions as well. However, I'll approve since this functionally does as it sets out to do.

**Review by thedadams — APPROVED** [review]

File: electron/main.mjs (hunk 2 above)
Comment (g-linville) [reply electron/main.mjs:34]: I suppose so, but right now the browser tool is the only tool that reads any kind of settings file for itself.

**Review by tylerslaton — APPROVED** [review]

**Review by thedadams — APPROVED** [review]

**Comment by g-linville** [comment]
Pushing a change dismissed the previous approvals for some reason, so reviews re-requested.

**Comment by g-linville** [comment]
Wow that was fast lol

**Pull request merged**